{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "248c3ab5-8948-4733-8561-30c995fe1ab4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!pip install cyvcf2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3459363b-0122-49e6-8f07-9afb83e2d83e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from cyvcf2 import VCF\n",
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4a05be6a-3e86-4aa6-9dc1-cbc1abd3a9bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "ann_fields = [\n",
    "    \"Allele\", \"Annotation\", \"Annotation_Impact\", \"Gene_Name\",\n",
    "    \"Gene_ID\", \"Feature_Type\", \"Feature_ID\", \"Transcript_Biotype\",\n",
    "    \"Rank\", \"HGVS.c\", \"HGVS.p\", \"cDNA.pos\", \"CDS.pos\", \"AA.pos\",\n",
    "    \"Distance\", \"Errors_Warnings_Info\"\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d700060e-a1db-44bf-a53e-327fb7c785ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_snpEff_ann(vcf_path):\n",
    "    vcf = VCF(vcf_path)\n",
    "    rows = []\n",
    "\n",
    "    for var in vcf:\n",
    "        chrom = var.CHROM\n",
    "        pos = var.POS\n",
    "        ref = var.REF\n",
    "        alt = var.ALT[0]\n",
    "        qual = var.QUAL\n",
    "\n",
    "        info_dp = var.INFO.get(\"DP\")\n",
    "\n",
    "        anns = var.INFO.get(\"ANN\")\n",
    "        if anns is None:\n",
    "            continue\n",
    "\n",
    "        # ANN can have multiple comma-separated annotations\n",
    "        for ann in anns.split(\",\"):\n",
    "            fields = ann.split(\"|\")\n",
    "            fields = fields + [\"\"] * (len(ann_fields) - len(fields))\n",
    "\n",
    "            row = {\n",
    "                \"CHROM\": chrom,\n",
    "                \"POS\": pos,\n",
    "                \"REF\": ref,\n",
    "                \"ALT\": alt,\n",
    "                \"QUAL\": qual,\n",
    "                \"DP\": info_dp,\n",
    "            }\n",
    "\n",
    "            for k, v in zip(ann_fields, fields):\n",
    "                row[k] = v\n",
    "\n",
    "            rows.append(row)\n",
    "\n",
    "    return pd.DataFrame(rows)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4d5fbe14-68f5-458c-aa53-051b303c3d35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>DP</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Annotation_Impact</th>\n",
       "      <th>Gene_Name</th>\n",
       "      <th>...</th>\n",
       "      <th>Feature_ID</th>\n",
       "      <th>Transcript_Biotype</th>\n",
       "      <th>Rank</th>\n",
       "      <th>HGVS.c</th>\n",
       "      <th>HGVS.p</th>\n",
       "      <th>cDNA.pos</th>\n",
       "      <th>CDS.pos</th>\n",
       "      <th>AA.pos</th>\n",
       "      <th>Distance</th>\n",
       "      <th>Errors_Warnings_Info</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Contig1453</td>\n",
       "      <td>127</td>\n",
       "      <td>T</td>\n",
       "      <td>TGG</td>\n",
       "      <td>170.490005</td>\n",
       "      <td>13</td>\n",
       "      <td>TGG</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Bma000001.1</td>\n",
       "      <td>...</td>\n",
       "      <td>Bma000001.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1/1</td>\n",
       "      <td>c.127_128dupGG</td>\n",
       "      <td>p.Asn44fs</td>\n",
       "      <td>129/954</td>\n",
       "      <td>129/954</td>\n",
       "      <td>43/317</td>\n",
       "      <td></td>\n",
       "      <td>WARNING_TRANSCRIPT_NO_START_CODON&amp;INFO_REALIGN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Contig1453</td>\n",
       "      <td>127</td>\n",
       "      <td>T</td>\n",
       "      <td>TGG</td>\n",
       "      <td>170.490005</td>\n",
       "      <td>13</td>\n",
       "      <td>TGG</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>Bma000002.1</td>\n",
       "      <td>...</td>\n",
       "      <td>Bma000002.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td></td>\n",
       "      <td>c.*1976_*1977dupCC</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1977</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Contig1453</td>\n",
       "      <td>127</td>\n",
       "      <td>T</td>\n",
       "      <td>TGG</td>\n",
       "      <td>170.490005</td>\n",
       "      <td>13</td>\n",
       "      <td>TGG</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>Bma000003.1</td>\n",
       "      <td>...</td>\n",
       "      <td>Bma000003.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td></td>\n",
       "      <td>c.*3531_*3532dupCC</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>3532</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Contig1453</td>\n",
       "      <td>143</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>218.990005</td>\n",
       "      <td>15</td>\n",
       "      <td>C</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>LOW</td>\n",
       "      <td>Bma000001.1</td>\n",
       "      <td>...</td>\n",
       "      <td>Bma000001.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1/1</td>\n",
       "      <td>c.141A&gt;C</td>\n",
       "      <td>p.Thr47Thr</td>\n",
       "      <td>141/954</td>\n",
       "      <td>141/954</td>\n",
       "      <td>47/317</td>\n",
       "      <td></td>\n",
       "      <td>WARNING_TRANSCRIPT_NO_START_CODON</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Contig1453</td>\n",
       "      <td>143</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>218.990005</td>\n",
       "      <td>15</td>\n",
       "      <td>C</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>Bma000002.1</td>\n",
       "      <td>...</td>\n",
       "      <td>Bma000002.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td></td>\n",
       "      <td>c.*1962T&gt;G</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1962</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        CHROM  POS REF  ALT        QUAL  DP Allele               Annotation  \\\n",
       "0  Contig1453  127   T  TGG  170.490005  13    TGG       frameshift_variant   \n",
       "1  Contig1453  127   T  TGG  170.490005  13    TGG  downstream_gene_variant   \n",
       "2  Contig1453  127   T  TGG  170.490005  13    TGG  downstream_gene_variant   \n",
       "3  Contig1453  143   A    C  218.990005  15      C       synonymous_variant   \n",
       "4  Contig1453  143   A    C  218.990005  15      C  downstream_gene_variant   \n",
       "\n",
       "  Annotation_Impact    Gene_Name  ...   Feature_ID Transcript_Biotype Rank  \\\n",
       "0              HIGH  Bma000001.1  ...  Bma000001.1     protein_coding  1/1   \n",
       "1          MODIFIER  Bma000002.1  ...  Bma000002.1     protein_coding        \n",
       "2          MODIFIER  Bma000003.1  ...  Bma000003.1     protein_coding        \n",
       "3               LOW  Bma000001.1  ...  Bma000001.1     protein_coding  1/1   \n",
       "4          MODIFIER  Bma000002.1  ...  Bma000002.1     protein_coding        \n",
       "\n",
       "               HGVS.c      HGVS.p cDNA.pos  CDS.pos  AA.pos Distance  \\\n",
       "0      c.127_128dupGG   p.Asn44fs  129/954  129/954  43/317            \n",
       "1  c.*1976_*1977dupCC                                           1977   \n",
       "2  c.*3531_*3532dupCC                                           3532   \n",
       "3            c.141A>C  p.Thr47Thr  141/954  141/954  47/317            \n",
       "4          c.*1962T>G                                           1962   \n",
       "\n",
       "                                Errors_Warnings_Info  \n",
       "0  WARNING_TRANSCRIPT_NO_START_CODON&INFO_REALIGN...  \n",
       "1                                                     \n",
       "2                                                     \n",
       "3                  WARNING_TRANSCRIPT_NO_START_CODON  \n",
       "4                                                     \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = parse_snpEff_ann(\"all_annotated_sample.vcf\")\n",
    "df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e3b60b9c-a778-4970-9b77-a0f8571f28ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>DP</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Annotation_Impact</th>\n",
       "      <th>Gene_Name</th>\n",
       "      <th>...</th>\n",
       "      <th>Feature_ID</th>\n",
       "      <th>Transcript_Biotype</th>\n",
       "      <th>Rank</th>\n",
       "      <th>HGVS.c</th>\n",
       "      <th>HGVS.p</th>\n",
       "      <th>cDNA.pos</th>\n",
       "      <th>CDS.pos</th>\n",
       "      <th>AA.pos</th>\n",
       "      <th>Distance</th>\n",
       "      <th>Errors_Warnings_Info</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Contig1453</td>\n",
       "      <td>127</td>\n",
       "      <td>T</td>\n",
       "      <td>TGG</td>\n",
       "      <td>170.490005</td>\n",
       "      <td>13</td>\n",
       "      <td>TGG</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Bma000001.1</td>\n",
       "      <td>...</td>\n",
       "      <td>Bma000001.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1/1</td>\n",
       "      <td>c.127_128dupGG</td>\n",
       "      <td>p.Asn44fs</td>\n",
       "      <td>129/954</td>\n",
       "      <td>129/954</td>\n",
       "      <td>43/317</td>\n",
       "      <td></td>\n",
       "      <td>WARNING_TRANSCRIPT_NO_START_CODON&amp;INFO_REALIGN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Contig1453</td>\n",
       "      <td>217</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>288.549988</td>\n",
       "      <td>39</td>\n",
       "      <td>A</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>Bma000001.1</td>\n",
       "      <td>...</td>\n",
       "      <td>Bma000001.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1/1</td>\n",
       "      <td>c.215C&gt;A</td>\n",
       "      <td>p.Ala72Glu</td>\n",
       "      <td>215/954</td>\n",
       "      <td>215/954</td>\n",
       "      <td>72/317</td>\n",
       "      <td></td>\n",
       "      <td>WARNING_TRANSCRIPT_NO_START_CODON</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Contig1453</td>\n",
       "      <td>231</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>151.899994</td>\n",
       "      <td>60</td>\n",
       "      <td>T</td>\n",
       "      <td>stop_gained</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Bma000001.1</td>\n",
       "      <td>...</td>\n",
       "      <td>Bma000001.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1/1</td>\n",
       "      <td>c.229G&gt;T</td>\n",
       "      <td>p.Glu77*</td>\n",
       "      <td>229/954</td>\n",
       "      <td>229/954</td>\n",
       "      <td>77/317</td>\n",
       "      <td></td>\n",
       "      <td>WARNING_TRANSCRIPT_NO_START_CODON</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Contig1453</td>\n",
       "      <td>232</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>151.580002</td>\n",
       "      <td>61</td>\n",
       "      <td>C</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>Bma000001.1</td>\n",
       "      <td>...</td>\n",
       "      <td>Bma000001.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1/1</td>\n",
       "      <td>c.230A&gt;C</td>\n",
       "      <td>p.Glu77Ala</td>\n",
       "      <td>230/954</td>\n",
       "      <td>230/954</td>\n",
       "      <td>77/317</td>\n",
       "      <td></td>\n",
       "      <td>WARNING_TRANSCRIPT_NO_START_CODON</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Contig1453</td>\n",
       "      <td>236</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>305.779999</td>\n",
       "      <td>62</td>\n",
       "      <td>C</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>Bma000001.1</td>\n",
       "      <td>...</td>\n",
       "      <td>Bma000001.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1/1</td>\n",
       "      <td>c.234A&gt;C</td>\n",
       "      <td>p.Leu78Phe</td>\n",
       "      <td>234/954</td>\n",
       "      <td>234/954</td>\n",
       "      <td>78/317</td>\n",
       "      <td></td>\n",
       "      <td>WARNING_TRANSCRIPT_NO_START_CODON</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         CHROM  POS REF  ALT        QUAL  DP Allele          Annotation  \\\n",
       "0   Contig1453  127   T  TGG  170.490005  13    TGG  frameshift_variant   \n",
       "21  Contig1453  217   C    A  288.549988  39      A    missense_variant   \n",
       "33  Contig1453  231   G    T  151.899994  60      T         stop_gained   \n",
       "36  Contig1453  232   A    C  151.580002  61      C    missense_variant   \n",
       "39  Contig1453  236   A    C  305.779999  62      C    missense_variant   \n",
       "\n",
       "   Annotation_Impact    Gene_Name  ...   Feature_ID Transcript_Biotype Rank  \\\n",
       "0               HIGH  Bma000001.1  ...  Bma000001.1     protein_coding  1/1   \n",
       "21          MODERATE  Bma000001.1  ...  Bma000001.1     protein_coding  1/1   \n",
       "33              HIGH  Bma000001.1  ...  Bma000001.1     protein_coding  1/1   \n",
       "36          MODERATE  Bma000001.1  ...  Bma000001.1     protein_coding  1/1   \n",
       "39          MODERATE  Bma000001.1  ...  Bma000001.1     protein_coding  1/1   \n",
       "\n",
       "            HGVS.c      HGVS.p cDNA.pos  CDS.pos  AA.pos Distance  \\\n",
       "0   c.127_128dupGG   p.Asn44fs  129/954  129/954  43/317            \n",
       "21        c.215C>A  p.Ala72Glu  215/954  215/954  72/317            \n",
       "33        c.229G>T    p.Glu77*  229/954  229/954  77/317            \n",
       "36        c.230A>C  p.Glu77Ala  230/954  230/954  77/317            \n",
       "39        c.234A>C  p.Leu78Phe  234/954  234/954  78/317            \n",
       "\n",
       "                                 Errors_Warnings_Info  \n",
       "0   WARNING_TRANSCRIPT_NO_START_CODON&INFO_REALIGN...  \n",
       "21                  WARNING_TRANSCRIPT_NO_START_CODON  \n",
       "33                  WARNING_TRANSCRIPT_NO_START_CODON  \n",
       "36                  WARNING_TRANSCRIPT_NO_START_CODON  \n",
       "39                  WARNING_TRANSCRIPT_NO_START_CODON  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter impactful variants\n",
    "impact_df = df[df[\"Annotation_Impact\"].isin([\"MODERATE\", \"HIGH\"])]\n",
    "impact_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f6c25a6c-c559-491d-9a2c-5bdee2324f59",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Gene_ID\n",
       "null           75\n",
       "null.2         75\n",
       "null.3         75\n",
       "null-null.2    18\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Count mutations per gene\n",
    "df[\"Gene_ID\"].value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a02c6f6-8a1c-42cf-ab7b-a6b19b116bed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Annotation = functional effect (missense_variant, synonymous_variant, stop_gained, intron_variant, upstream_gene_variant).\n",
    "# Annotation_Impact= impact category (HIGH, MODERATE, LOW, MODIFIER)\n",
    "# Gene_name= might match GFF attributes (ex: Bma002878.1).\n",
    "# Gene_ID= gene identifier from GFF attributes.\n",
    "# Feature_Type= type of sequence affected (transcript, mRNA, CDS, exon)\n",
    "# Feature_ID=ID of transcript/feature affected (Bma002878.1).\n",
    "# Transcript_Biotype=class transcript GFF (protein_coding, lncRNA, pseudogene). \n",
    "# Rank=exons or transcripts (<exon_number>/<total_exons>, ex: the variant occurs in exon 3 of 7).\n",
    "# HGVS.c=HGVS notation at the coding DNA level.\n",
    "# HGVS.p=HGVS notation at the protein level.\n",
    "# cDNA.pos=Position in the cDNA sequence (including UTRs). Format: <pos>/<length>, Example: 678/1450\n",
    "# CDS.pos=Position within the coding sequence only.If the mutation is outside the CDS (e.g., UTR, intron), this is empty.\n",
    "# AA.pos=Position within the protein sequence. Empty for noncoding variants.\n",
    "# Distance=Distance to the nearest feature if the variant is not inside a gene/transcript.\n",
    "# Errors_Warnings_Info=Any warnings, errors, or special flags snpEff adds (WARNING_TRANSCRIPT_NO_START_CODON, WARNING_TRANSCRIPT_INCOMPLETE, ERROR_CHROMOSOME_NOT_FOUND, INFO_REALIGN_3_PRIME).\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1f78801-2e10-4c9f-b07c-92537a528201",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
