{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "248c3ab5-8948-4733-8561-30c995fe1ab4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!pip install cyvcf2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3459363b-0122-49e6-8f07-9afb83e2d83e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from cyvcf2 import VCF\n",
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4a05be6a-3e86-4aa6-9dc1-cbc1abd3a9bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "ann_fields = [\n",
    "    \"Allele\", \"Annotation\", \"Annotation_Impact\", \"Gene_Name\",\n",
    "    \"Gene_ID\", \"Feature_Type\", \"Feature_ID\", \"Transcript_Biotype\",\n",
    "    \"Rank\", \"HGVS.c\", \"HGVS.p\", \"cDNA.pos\", \"CDS.pos\", \"AA.pos\",\n",
    "    \"Distance\", \"Errors_Warnings_Info\"\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d700060e-a1db-44bf-a53e-327fb7c785ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_snpEff_ann(vcf_path):\n",
    "    vcf = VCF(vcf_path)\n",
    "    rows = []\n",
    "\n",
    "    for var in vcf:\n",
    "        chrom = var.CHROM\n",
    "        pos = var.POS\n",
    "        ref = var.REF\n",
    "        alt = var.ALT[0]\n",
    "        qual = var.QUAL\n",
    "\n",
    "        info_dp = var.INFO.get(\"DP\")\n",
    "\n",
    "        anns = var.INFO.get(\"ANN\")\n",
    "        if anns is None:\n",
    "            continue\n",
    "\n",
    "        # ANN can have multiple comma-separated annotations\n",
    "        for ann in anns.split(\",\"):\n",
    "            fields = ann.split(\"|\")\n",
    "            fields = fields + [\"\"] * (len(ann_fields) - len(fields))\n",
    "\n",
    "            row = {\n",
    "                \"CHROM\": chrom,\n",
    "                \"POS\": pos,\n",
    "                \"REF\": ref,\n",
    "                \"ALT\": alt,\n",
    "                \"QUAL\": qual,\n",
    "                \"DP\": info_dp,\n",
    "            }\n",
    "\n",
    "            for k, v in zip(ann_fields, fields):\n",
    "                row[k] = v\n",
    "\n",
    "            rows.append(row)\n",
    "\n",
    "    return pd.DataFrame(rows)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d5fbe14-68f5-458c-aa53-051b303c3d35",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = parse_snpEff_ann(\"all_annotated_sample.vcf\")\n",
    "df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3b60b9c-a778-4970-9b77-a0f8571f28ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter impactful variants\n",
    "impact_df = df[df[\"Annotation_Impact\"].isin([\"MODERATE\", \"HIGH\"])]\n",
    "impact_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6c25a6c-c559-491d-9a2c-5bdee2324f59",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Count mutations per gene\n",
    "df[\"Gene_ID\"].value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a02c6f6-8a1c-42cf-ab7b-a6b19b116bed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Annotation = functional effect (missense_variant, synonymous_variant, stop_gained, intron_variant, upstream_gene_variant).\n",
    "# Annotation_Impact= impact category (HIGH, MODERATE, LOW, MODIFIER)\n",
    "# Gene_name= might match GFF attributes (ex: Bma002878.1).\n",
    "# Gene_ID= gene identifier from GFF attributes.\n",
    "# Feature_Type= type of sequence affected (transcript, mRNA, CDS, exon)\n",
    "# Feature_ID=ID of transcript/feature affected (Bma002878.1).\n",
    "# Transcript_Biotype=class transcript GFF (protein_coding, lncRNA, pseudogene). \n",
    "# Rank=exons or transcripts (<exon_number>/<total_exons>, ex: the variant occurs in exon 3 of 7).\n",
    "# HGVS.c=HGVS notation at the coding DNA level.\n",
    "# HGVS.p=HGVS notation at the protein level.\n",
    "# cDNA.pos=Position in the cDNA sequence (including UTRs). Format: <pos>/<length>, Example: 678/1450\n",
    "# CDS.pos=Position within the coding sequence only.If the mutation is outside the CDS (e.g., UTR, intron), this is empty.\n",
    "# AA.pos=Position within the protein sequence. Empty for noncoding variants.\n",
    "# Distance=Distance to the nearest feature if the variant is not inside a gene/transcript.\n",
    "# Errors_Warnings_Info=Any warnings, errors, or special flags snpEff adds (WARNING_TRANSCRIPT_NO_START_CODON, WARNING_TRANSCRIPT_INCOMPLETE, ERROR_CHROMOSOME_NOT_FOUND, INFO_REALIGN_3_PRIME).\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1f78801-2e10-4c9f-b07c-92537a528201",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
