# Bmas GERP pipeline

### 1. Purpose

The previous experiment with alignment of the reads on baits allowed us to investigate the intersection of conserved regions (HyPhy) and regions under selections (GERP). Unfortunately, the SIFT annotations that could pinpoint the deleterious mutations were only made possible by the B.mas annotated genome. And blasting the baits on the B.mas genome gave uncertain results.
Futhermore we developped a method of hard filtering the GERP score (GERP>4 selected) that showed potential deleterious alleles, but did not enabled to vizualize high-scoring regions.

Therefore, reads have been now directly aligned to the B.mas genome to have a direct read on the GERP score and annotations. The SIFT analysis will not be undertaken as it gives dispensable results.

### 2. Alignement 

As we work on a chromosome-scale genome and will make mostly use of SNP, in plus of using VCF/ROH data later, we will not use SPADES to align the reads. Instead I am using the ROH pipeline, with array scripts to process the data quicker. 

Then we will use the VCF and the reference genome with the 3 scripts found here:
/home/tmichel/scratch/Deleterious_alleles_PNG_Bmas/MSA
This pipeline provides the MSA files for each genes.

### 3.Filtering

As mentioned in Hannah's thesis, two filtering streps can be found below, filtering MSA with 70% missing data and 75% missing samples.
/home/tmichel/scratch/Deleterious_alleles_PNG_baits/CAPTUS/05_phylogeny_FastTreeAstral

### 4.phylogenetic reconstruction

As mentioned in Hannah's thesis, we will reconstruct the genes and species trees with respectively Fasttree and Astral.

~/scratch/Deleterious_alleles_PNG_baits/CAPTUS/05_phylogeny_FastTreeAstral
