{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1f16d81e-4627-4bb2-a6d9-051888e6cbf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13886525-b4f1-4fe9-8f84-3fd57f31186f",
   "metadata": {},
   "source": [
    "## 1. conserved regions on B. masoniana chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ee5c568a-ca6a-445d-9d9a-f3b7c36b8ccf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                Contig  Start  End  - -- ---\n",
      "0   ACmerged_contig_66    358  363  .  .   .\n",
      "1   ACmerged_contig_66    367  369  .  .   .\n",
      "2   ACmerged_contig_66    373  375  .  .   .\n",
      "3  ACmerged_contig_694    175  183  .  .   .\n",
      "4  ACmerged_contig_806    907  909  .  .   .\n"
     ]
    }
   ],
   "source": [
    "# Load the BED file with the conserved regions\n",
    "df = pd.read_csv('/home/thibauld/Documents/Bioinformatics/Deleterious_alleles_pipeline/Deleterious_alleles_PNG/intersection_GERP_HyPhy_ASTRAL/intersection_GERP_HyPhy.bed', \n",
    "                 sep='\\t', \n",
    "                 header=None, \n",
    "                 names=['Contig', 'Start', 'End', '-', '--', '---'])\n",
    "\n",
    "# Display the first few rows\n",
    "print(df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfe2dbc3-7e47-45cc-816d-a9ec7662728e",
   "metadata": {},
   "outputs": [],
   "source": [
    "/home/thibauld/Documents/Bioinformatics/Deleterious_alleles_pipeline/Deleterious_alleles_PNG/blast_masoniana/results_blast_best_per_cluster_Only_one_match.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "03ade3b9-791e-46b5-bbd9-e4846864f326",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   cluster_id                                      qseqid      sseqid  pident  \\\n",
      "0           1  Hannah_Begonia_baits-ACmerged_contig_10054  scaffold11   100.0   \n",
      "1           2   Hannah_Begonia_baits-ACmerged_contig_1011   scaffold1   100.0   \n",
      "2           3  Hannah_Begonia_baits-ACmerged_contig_10412  scaffold14   100.0   \n",
      "3           6  Hannah_Begonia_baits-ACmerged_contig_10858  scaffold12   100.0   \n",
      "4           9  Hannah_Begonia_baits-ACmerged_contig_10984   scaffold8   100.0   \n",
      "\n",
      "   length    sstart      send      slen         evalue  bitscore  \n",
      "0      63  43902442  43902504  46644841   7.150000e-24       117  \n",
      "1     272   6657182   6656911  84930515  2.870000e-140       503  \n",
      "2      84  35880662  35880579  45922387   8.590000e-36       156  \n",
      "3     399   7226210   7226608  46145620   0.000000e+00       737  \n",
      "4     172  46732024  46732195  51801790   1.540000e-84       318  \n"
     ]
    }
   ],
   "source": [
    "# Load the BED file with the position of the baits on the scaffolds.\n",
    "df = pd.read_csv('/home/thibauld/Documents/Bioinformatics/Deleterious_alleles_pipeline/Deleterious_alleles_PNG/blast_masoniana/results_blast_best_per_cluster_Only_one_match.csv', \n",
    "                 sep=',')\n",
    "\n",
    "# Display the first few rows\n",
    "print(df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d9839de-c841-4ee3-ad53-d4a1e4685e61",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
