{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b1d69216-2cff-4915-9048-a27de4215f5a",
   "metadata": {},
   "source": [
    "# Load SIFT-annotated VCF and filter variants to keep deleterious with high effects  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a33f78a3-5886-49ee-ba38-6ef4f38a0119",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Annotations impacts definitions\n",
    "# https://pcingola.github.io/SnpEff/snpeff/inputoutput/\n",
    "\n",
    "# How many HIGH, MODERATE, LOW, MODIFIER variants in our set of data?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d2c78c07-3ef8-4d25-a481-1244338dada7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load part of the modules\n",
    "import sys\n",
    "from cyvcf2 import VCF\n",
    "import csv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1419b249-4798-48e2-b68a-586deff2116e",
   "metadata": {},
   "source": [
    "## 1. Load the VCF file in Python dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4fb3d461-9e23-4a5c-b4d2-ae6cbe1d7bbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For memory, headers are:\n",
    "\n",
    "#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  B.S.Pet.102     B.S.Pet.104     B.S.Pet.105     B.S.Pet.127     B.S.Pet.128     B.S.Pet.129   B.S.Pet.156     B.S.Pet.157     B.S.Pet.235     B.S.Pet.239     B.S.Pet.241     B.S.Pet.242     B.S.Pet.243     B.S.Pet.244     B.S.Pet.245     B.S.Pet.60    B.S.Pet.60a     B.S.Pet.78      B.S.Pet.80      B.S.Pet.83      B.S.Pet.84      B.S.Pet.89      B.S.Pet.93      B.S.Pet.96      B.S.Pet.97      B.S.Pet.98    B.S.Pet.99      B.S.Pet.B1.133  B.S.Pet.B1.39   B.S.Pet.B13.111 B.S.Pet.B13.44  B.S.Pet.B15.110 B.S.Pet.B15.40  B.S.Pet.B33.131 B.S.Pet.B33.43  B.S.Pet.B44.132       B.S.Pet.B44.41  B.S.Pet.B46.112 B.S.Pet.B46.42  B.S.Pet.ELAE1.195       B.S.Pet.ELAE104.228     B.S.Pet.ELAE111.230     B.S.Pet.ELAE118.233           B.S.Pet.ELAE119.234     B.S.Pet.ELAE27.204      B.S.Pet.ELAE4.196       B.S.Pet.ELAE52.211      B.S.Pet.ELAE62.213      B.S.Pet.ELAE72.214      B.S.Pet.ELAE81.215    B.S.Pet.ELAE82.216      B.S.Pet.ELAE83.217      B.S.Pet.ELAE86.218      B.S.Pet.ELAE93.222      B.S.Pet.ELAE94.223      B.S.Pet.ELAE98.225            B.S.Sym.117     B.S.Sym.117a    B.S.Sym.28      B.S.Sym.28a     B.S.Sym.32      B.S.Sym.48      B.S.Sym.48a     B.S.Sym.B32.140 B.S.Sym.B34.121 B.S.Sym.ELAE46.209    B.aff.augustae30        B.aff.augustae56        B.aff.jagorii65 B.aff.rubella169        B.aff.symbracteosa35    B.aff.tafaensis87       B.aff.weigallii92     B.aikrono237    B.aikrono240    B.antsiranensis175      B.arfakensis29  B.asaroensis115 B.asaroensis46  B.asaroensis46a B.atricha49     B.atricha52   B.augustae164   B.bipinnatifida1        B.bipinnatifida178      B.bipinnatifida178a     B.bipinnatifida1a       B.bipinnatifida38       B.bipinnatifida58     B.brachybotrys147       B.brachybotrys147a      B.brachybotrys205       B.brachybotrys205a      B.bracteosa21   B.brassii202    B.brassii202a   B.brassii207  B.brassii219    B.brassii227    B.cf.fulvovillosa31     B.cf.fulvovillosa31a    B.cf.strigosa120        B.cf.strigosa142        B.chambersiae220              B.cyrtandroides.sp.nov14        B.cyrtandroides.sp.nov14a       B.cyrtandroides.sp.nov16        B.cyrtandroides.sp.nov16a       B.cyrtandroides.sp.nov17      B.cyrtandroides.sp.nov19        B.erodifolia45  B.erodifolia45a B.fractalifolia236      B.fulvovillosa151       B.fulvovillosa151a      B.kaniensis108        B.kaniensis125  B.kaniensis153  B.kaniensis153a B.kaniensis153b B.kaniensis161  B.kaniensis201  B.kaniensis212  B.kaniensis221  B.kaniensis224  B.kaniensis231        B.kemumuensis27 B.lagunensis64  B.leucostricta176       B.longifolia34  B.maguniana180  B.maguniana2    B.maguniana206  B.maguniana226  B.maguniana229        B.maguniana2a   B.maguniana61   B.magunianaHOLO198      B.magxbrassii15 B.malachosticta68       B.maynensis37   B.mimikaensis159        B.mimikaensis159a     B.mindanaensis74        B.modestiflora168       B.octopetala11  B.oligandra162  B.oligandra200  B.pendulousOligandrae13 B.pentandra210  B.peruviana36 B.sandaunensis238       B.sandsiana203  B.serratipetalaELAE115.232      B.sp.ABEG248.174        B.sp.CKYang1598.179     B.sp.EKBOE46.171        B.sp.P.Blanc172       B.sp.P.Blanc2173        B.sp.Rubite488.177      B.sp.SUBOE60.26 B.sp.SUBOE81.54 B.sp.Wilkie29116.71     B.sp.Wilkie29124.76     B.sp.Wilkie29127.72   B.speluncae62   B.stenogyna67   B.stictopoda170 B.stictopoda51  B.stilandra18   B.stilandra18a  B.strigosa113   B.strigosa116   B.subciliata24  B.symbracteosa158     B.symsanguinea139       B.symsanguinea25        B.symsanguinea25a       B.trichopoda50  B.trichopoda53  B.uniflora167   B.vinkii134     B.vinkii138   B.vinkii197     B.vinkii199     B.vinkii208     B.vinkii3       B.vinkii3a      B.vinkii3b      B.weigallii57\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "aed0678e-1236-406e-9e00-3635597bfc1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a simple script to extract this fields (without separating them at first).\n",
    "import pandas as pd\n",
    "import gzip\n",
    "\n",
    "def read_vcf_to_dataframe(vcf_file):\n",
    "    opener = gzip.open if vcf_file.endswith(\".gz\") else open\n",
    "\n",
    "    with opener(vcf_file, \"rt\") as f:\n",
    "        for line in f:\n",
    "            if line.startswith(\"#CHROM\"):\n",
    "                header = line.strip().lstrip(\"#\").split(\"\\t\")\n",
    "                break\n",
    "\n",
    "        rows = []\n",
    "        for line in f:\n",
    "            if line.startswith(\"#\"):\n",
    "                continue\n",
    "            rows.append(line.strip().split(\"\\t\"))\n",
    "\n",
    "    df = pd.DataFrame(rows, columns=header)\n",
    "\n",
    "    # Basic type conversion\n",
    "    df[\"POS\"] = df[\"POS\"].astype(int)\n",
    "    df[\"QUAL\"] = pd.to_numeric(df[\"QUAL\"], errors=\"coerce\")\n",
    "\n",
    "    # ------------------\n",
    "    # Parse INFO field\n",
    "    # ------------------\n",
    "    info_dicts = []\n",
    "    for info in df[\"INFO\"]:\n",
    "        d = {}\n",
    "        for item in info.split(\";\"):\n",
    "            if \"=\" in item:\n",
    "                k, v = item.split(\"=\", 1)\n",
    "                d[k] = v\n",
    "            else:\n",
    "                d[item] = True\n",
    "        info_dicts.append(d)\n",
    "\n",
    "    info_df = pd.DataFrame(info_dicts)\n",
    "\n",
    "    # ------------------\n",
    "    # Parse samples (keep one column per sample)\n",
    "    # ------------------\n",
    "    sample_cols = header[9:]\n",
    "\n",
    "    def parse_sample(sample_str, keys):\n",
    "        values = sample_str.split(\":\")\n",
    "        return dict(zip(keys, values))\n",
    "\n",
    "    format_keys = df[\"FORMAT\"].iloc[0].split(\":\")\n",
    "\n",
    "    samples_df = pd.DataFrame({\n",
    "        sample: df[sample].apply(lambda x: parse_sample(x, format_keys))\n",
    "        for sample in sample_cols\n",
    "    })\n",
    "\n",
    "    # ------------------\n",
    "    # Final dataframe\n",
    "    # ------------------\n",
    "    fixed_cols = [\"CHROM\", \"POS\", \"ID\", \"REF\", \"ALT\", \"QUAL\", \"FILTER\"]\n",
    "\n",
    "    final_df = pd.concat(\n",
    "        [\n",
    "            df[fixed_cols],\n",
    "            info_df,\n",
    "            samples_df\n",
    "        ],\n",
    "        axis=1\n",
    "    )\n",
    "\n",
    "    return final_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ce221ee2-e009-43eb-9171-65ae261cc1e6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(422, 215)\n",
      "['CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'AC', 'AF', 'AN', 'BaseQRankSum', 'DP', 'ExcessHet', 'FS', 'InbreedingCoeff', 'MLEAC', 'MLEAF', 'MQ', 'MQRankSum', 'QD']\n"
     ]
    }
   ],
   "source": [
    "# Testing the extraction script\n",
    "\n",
    "vcf = \"/home/thibauld/Documents/Bioinformatics/Deleterious_alleles_pipeline/Deleterious_alleles_PNG/conserved_SNPs/filtered_in_regions.vcf\"\n",
    "df = read_vcf_to_dataframe(vcf)\n",
    "\n",
    "print(df.shape)\n",
    "print(df.columns.tolist()[:20])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9c56aef1-03b1-483b-beba-a933ae22e0fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>AC</th>\n",
       "      <th>AF</th>\n",
       "      <th>AN</th>\n",
       "      <th>...</th>\n",
       "      <th>B.uniflora167</th>\n",
       "      <th>B.vinkii134</th>\n",
       "      <th>B.vinkii138</th>\n",
       "      <th>B.vinkii197</th>\n",
       "      <th>B.vinkii199</th>\n",
       "      <th>B.vinkii208</th>\n",
       "      <th>B.vinkii3</th>\n",
       "      <th>B.vinkii3a</th>\n",
       "      <th>B.vinkii3b</th>\n",
       "      <th>B.weigallii57</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>scaffold11</td>\n",
       "      <td>2133652</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>C,GTCA,A</td>\n",
       "      <td>2476.74</td>\n",
       "      <td>.</td>\n",
       "      <td>2,1,1</td>\n",
       "      <td>0.005263,0.002632,0.002632</td>\n",
       "      <td>380</td>\n",
       "      <td>...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '10,0,0,0', 'DP': '10', 'G...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0,0', 'DP': '0', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0,0', 'DP': '0', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0,0', 'DP': '0', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0,0', 'DP': '0', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0,0', 'DP': '0', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0,0', 'DP': '0', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '3,0,0,0', 'DP': '3', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '2,0,0,0', 'DP': '2', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0,0', 'DP': '0', 'GQ'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>scaffold11</td>\n",
       "      <td>5971895</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>5133.71</td>\n",
       "      <td>.</td>\n",
       "      <td>1</td>\n",
       "      <td>0.002632</td>\n",
       "      <td>380</td>\n",
       "      <td>...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '637,0', 'DP': '637', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0', 'DP': '0', 'GQ': '0...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '18,0', 'DP': '18', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '36,0', 'DP': '36', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '37,0', 'DP': '37', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '38,0', 'DP': '38', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '42,0', 'DP': '42', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '266,0', 'DP': '266', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '155,0', 'DP': '155', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '73,0', 'DP': '73', 'GQ': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>scaffold11</td>\n",
       "      <td>6230609</td>\n",
       "      <td>.</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>5654.54</td>\n",
       "      <td>.</td>\n",
       "      <td>1</td>\n",
       "      <td>0.002632</td>\n",
       "      <td>380</td>\n",
       "      <td>...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '739,0', 'DP': '739', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0', 'DP': '0', 'GQ': '0...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '110,0', 'DP': '110', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '48,0', 'DP': '48', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '23,0', 'DP': '23', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '30,0', 'DP': '30', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0', 'DP': '0', 'GQ': '0...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '349,0', 'DP': '349', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '265,0', 'DP': '265', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '92,0', 'DP': '92', 'GQ': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>scaffold11</td>\n",
       "      <td>6230610</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>2356.86</td>\n",
       "      <td>.</td>\n",
       "      <td>2</td>\n",
       "      <td>0.005263</td>\n",
       "      <td>380</td>\n",
       "      <td>...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '739,0', 'DP': '739', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0', 'DP': '0', 'GQ': '0...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '110,0', 'DP': '110', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '48,0', 'DP': '48', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '20,0', 'DP': '20', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '30,0', 'DP': '30', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0', 'DP': '0', 'GQ': '0...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '349,0', 'DP': '349', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '265,0', 'DP': '265', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '92,0', 'DP': '92', 'GQ': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>scaffold11</td>\n",
       "      <td>6230612</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>GACCTAC,A</td>\n",
       "      <td>69763.90</td>\n",
       "      <td>.</td>\n",
       "      <td>1,5</td>\n",
       "      <td>0.002632,0.013</td>\n",
       "      <td>380</td>\n",
       "      <td>...</td>\n",
       "      <td>{'GT': '0/2', 'AD': '359,0,453', 'DP': '812', ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '110,0,0', 'DP': '110', 'G...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '48,0,0', 'DP': '48', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '20,0,0', 'DP': '20', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '30,0,0', 'DP': '30', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '349,0,0', 'DP': '349', 'G...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '265,0,0', 'DP': '265', 'G...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '92,0,0', 'DP': '92', 'GQ'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>417</th>\n",
       "      <td>scaffold3</td>\n",
       "      <td>4691596</td>\n",
       "      <td>.</td>\n",
       "      <td>TATC</td>\n",
       "      <td>T</td>\n",
       "      <td>327.03</td>\n",
       "      <td>.</td>\n",
       "      <td>1</td>\n",
       "      <td>0.002632</td>\n",
       "      <td>380</td>\n",
       "      <td>...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '59,0', 'DP': '59', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0', 'DP': '0', 'GQ': '0...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0', 'DP': '0', 'GQ': '0...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0', 'DP': '0', 'GQ': '0...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0', 'DP': '0', 'GQ': '0...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0', 'DP': '0', 'GQ': '0...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0', 'DP': '0', 'GQ': '0...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '17,0', 'DP': '17', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '2,0', 'DP': '2', 'GQ': '6...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0', 'DP': '0', 'GQ': '0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>418</th>\n",
       "      <td>scaffold3</td>\n",
       "      <td>4691597</td>\n",
       "      <td>.</td>\n",
       "      <td>ATCTCAAAACATGAAGACATG</td>\n",
       "      <td>A,*,ACTCAAAACATGAAGACATG</td>\n",
       "      <td>811.50</td>\n",
       "      <td>.</td>\n",
       "      <td>1,1,2</td>\n",
       "      <td>0.002632,0.002632,0.005263</td>\n",
       "      <td>380</td>\n",
       "      <td>...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '59,0,0,0', 'DP': '59', 'G...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0,0', 'DP': '0', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0,0', 'DP': '0', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0,0', 'DP': '0', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0,0', 'DP': '0', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0,0', 'DP': '0', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0,0', 'DP': '0', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '17,0,0,0', 'DP': '17', 'G...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '2,0,0,0', 'DP': '2', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0,0', 'DP': '0', 'GQ'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419</th>\n",
       "      <td>scaffold3</td>\n",
       "      <td>4691599</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>*,T</td>\n",
       "      <td>524.95</td>\n",
       "      <td>.</td>\n",
       "      <td>2,1</td>\n",
       "      <td>0.005263,0.002632</td>\n",
       "      <td>380</td>\n",
       "      <td>...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '59,0,0', 'DP': '59', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '17,0,0', 'DP': '17', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '2,0,0', 'DP': '2', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>420</th>\n",
       "      <td>scaffold3</td>\n",
       "      <td>6043092</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>G,A</td>\n",
       "      <td>3726.32</td>\n",
       "      <td>.</td>\n",
       "      <td>1,2</td>\n",
       "      <td>0.002632,0.005263</td>\n",
       "      <td>380</td>\n",
       "      <td>...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '484,0,0', 'DP': '484', 'G...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '15,0,0', 'DP': '15', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '40,0,0', 'DP': '40', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '35,0,0', 'DP': '35', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '10,0,0', 'DP': '10', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '35,0,0', 'DP': '35', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '50,0,0', 'DP': '50', 'GQ'...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '269,0,0', 'DP': '269', 'G...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '200,0,0', 'DP': '200', 'G...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '106,0,0', 'DP': '106', 'G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421</th>\n",
       "      <td>scaffold15</td>\n",
       "      <td>42144934</td>\n",
       "      <td>.</td>\n",
       "      <td>AG</td>\n",
       "      <td>A,CG</td>\n",
       "      <td>239.80</td>\n",
       "      <td>.</td>\n",
       "      <td>1,2</td>\n",
       "      <td>0.002646,0.005291</td>\n",
       "      <td>378</td>\n",
       "      <td>...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '1,0,0', 'DP': '1', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...</td>\n",
       "      <td>{'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>422 rows × 215 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM       POS ID                    REF                       ALT  \\\n",
       "0    scaffold11   2133652  .                      G                  C,GTCA,A   \n",
       "1    scaffold11   5971895  .                      T                         A   \n",
       "2    scaffold11   6230609  .                      A                         C   \n",
       "3    scaffold11   6230610  .                      T                         C   \n",
       "4    scaffold11   6230612  .                      G                 GACCTAC,A   \n",
       "..          ...       ... ..                    ...                       ...   \n",
       "417   scaffold3   4691596  .                   TATC                         T   \n",
       "418   scaffold3   4691597  .  ATCTCAAAACATGAAGACATG  A,*,ACTCAAAACATGAAGACATG   \n",
       "419   scaffold3   4691599  .                      C                       *,T   \n",
       "420   scaffold3   6043092  .                      T                       G,A   \n",
       "421  scaffold15  42144934  .                     AG                      A,CG   \n",
       "\n",
       "         QUAL FILTER     AC                          AF   AN  ...  \\\n",
       "0     2476.74      .  2,1,1  0.005263,0.002632,0.002632  380  ...   \n",
       "1     5133.71      .      1                    0.002632  380  ...   \n",
       "2     5654.54      .      1                    0.002632  380  ...   \n",
       "3     2356.86      .      2                    0.005263  380  ...   \n",
       "4    69763.90      .    1,5              0.002632,0.013  380  ...   \n",
       "..        ...    ...    ...                         ...  ...  ...   \n",
       "417    327.03      .      1                    0.002632  380  ...   \n",
       "418    811.50      .  1,1,2  0.002632,0.002632,0.005263  380  ...   \n",
       "419    524.95      .    2,1           0.005263,0.002632  380  ...   \n",
       "420   3726.32      .    1,2           0.002632,0.005263  380  ...   \n",
       "421    239.80      .    1,2           0.002646,0.005291  378  ...   \n",
       "\n",
       "                                         B.uniflora167  \\\n",
       "0    {'GT': '0/0', 'AD': '10,0,0,0', 'DP': '10', 'G...   \n",
       "1    {'GT': '0/0', 'AD': '637,0', 'DP': '637', 'GQ'...   \n",
       "2    {'GT': '0/0', 'AD': '739,0', 'DP': '739', 'GQ'...   \n",
       "3    {'GT': '0/0', 'AD': '739,0', 'DP': '739', 'GQ'...   \n",
       "4    {'GT': '0/2', 'AD': '359,0,453', 'DP': '812', ...   \n",
       "..                                                 ...   \n",
       "417  {'GT': '0/0', 'AD': '59,0', 'DP': '59', 'GQ': ...   \n",
       "418  {'GT': '0/0', 'AD': '59,0,0,0', 'DP': '59', 'G...   \n",
       "419  {'GT': '0/0', 'AD': '59,0,0', 'DP': '59', 'GQ'...   \n",
       "420  {'GT': '0/0', 'AD': '484,0,0', 'DP': '484', 'G...   \n",
       "421  {'GT': '0/0', 'AD': '1,0,0', 'DP': '1', 'GQ': ...   \n",
       "\n",
       "                                           B.vinkii134  \\\n",
       "0    {'GT': '0/0', 'AD': '0,0,0,0', 'DP': '0', 'GQ'...   \n",
       "1    {'GT': '0/0', 'AD': '0,0', 'DP': '0', 'GQ': '0...   \n",
       "2    {'GT': '0/0', 'AD': '0,0', 'DP': '0', 'GQ': '0...   \n",
       "3    {'GT': '0/0', 'AD': '0,0', 'DP': '0', 'GQ': '0...   \n",
       "4    {'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...   \n",
       "..                                                 ...   \n",
       "417  {'GT': '0/0', 'AD': '0,0', 'DP': '0', 'GQ': '0...   \n",
       "418  {'GT': '0/0', 'AD': '0,0,0,0', 'DP': '0', 'GQ'...   \n",
       "419  {'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...   \n",
       "420  {'GT': '0/0', 'AD': '15,0,0', 'DP': '15', 'GQ'...   \n",
       "421  {'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...   \n",
       "\n",
       "                                           B.vinkii138  \\\n",
       "0    {'GT': '0/0', 'AD': '0,0,0,0', 'DP': '0', 'GQ'...   \n",
       "1    {'GT': '0/0', 'AD': '18,0', 'DP': '18', 'GQ': ...   \n",
       "2    {'GT': '0/0', 'AD': '110,0', 'DP': '110', 'GQ'...   \n",
       "3    {'GT': '0/0', 'AD': '110,0', 'DP': '110', 'GQ'...   \n",
       "4    {'GT': '0/0', 'AD': '110,0,0', 'DP': '110', 'G...   \n",
       "..                                                 ...   \n",
       "417  {'GT': '0/0', 'AD': '0,0', 'DP': '0', 'GQ': '0...   \n",
       "418  {'GT': '0/0', 'AD': '0,0,0,0', 'DP': '0', 'GQ'...   \n",
       "419  {'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...   \n",
       "420  {'GT': '0/0', 'AD': '40,0,0', 'DP': '40', 'GQ'...   \n",
       "421  {'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...   \n",
       "\n",
       "                                           B.vinkii197  \\\n",
       "0    {'GT': '0/0', 'AD': '0,0,0,0', 'DP': '0', 'GQ'...   \n",
       "1    {'GT': '0/0', 'AD': '36,0', 'DP': '36', 'GQ': ...   \n",
       "2    {'GT': '0/0', 'AD': '48,0', 'DP': '48', 'GQ': ...   \n",
       "3    {'GT': '0/0', 'AD': '48,0', 'DP': '48', 'GQ': ...   \n",
       "4    {'GT': '0/0', 'AD': '48,0,0', 'DP': '48', 'GQ'...   \n",
       "..                                                 ...   \n",
       "417  {'GT': '0/0', 'AD': '0,0', 'DP': '0', 'GQ': '0...   \n",
       "418  {'GT': '0/0', 'AD': '0,0,0,0', 'DP': '0', 'GQ'...   \n",
       "419  {'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...   \n",
       "420  {'GT': '0/0', 'AD': '35,0,0', 'DP': '35', 'GQ'...   \n",
       "421  {'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...   \n",
       "\n",
       "                                           B.vinkii199  \\\n",
       "0    {'GT': '0/0', 'AD': '0,0,0,0', 'DP': '0', 'GQ'...   \n",
       "1    {'GT': '0/0', 'AD': '37,0', 'DP': '37', 'GQ': ...   \n",
       "2    {'GT': '0/0', 'AD': '23,0', 'DP': '23', 'GQ': ...   \n",
       "3    {'GT': '0/0', 'AD': '20,0', 'DP': '20', 'GQ': ...   \n",
       "4    {'GT': '0/0', 'AD': '20,0,0', 'DP': '20', 'GQ'...   \n",
       "..                                                 ...   \n",
       "417  {'GT': '0/0', 'AD': '0,0', 'DP': '0', 'GQ': '0...   \n",
       "418  {'GT': '0/0', 'AD': '0,0,0,0', 'DP': '0', 'GQ'...   \n",
       "419  {'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...   \n",
       "420  {'GT': '0/0', 'AD': '10,0,0', 'DP': '10', 'GQ'...   \n",
       "421  {'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...   \n",
       "\n",
       "                                           B.vinkii208  \\\n",
       "0    {'GT': '0/0', 'AD': '0,0,0,0', 'DP': '0', 'GQ'...   \n",
       "1    {'GT': '0/0', 'AD': '38,0', 'DP': '38', 'GQ': ...   \n",
       "2    {'GT': '0/0', 'AD': '30,0', 'DP': '30', 'GQ': ...   \n",
       "3    {'GT': '0/0', 'AD': '30,0', 'DP': '30', 'GQ': ...   \n",
       "4    {'GT': '0/0', 'AD': '30,0,0', 'DP': '30', 'GQ'...   \n",
       "..                                                 ...   \n",
       "417  {'GT': '0/0', 'AD': '0,0', 'DP': '0', 'GQ': '0...   \n",
       "418  {'GT': '0/0', 'AD': '0,0,0,0', 'DP': '0', 'GQ'...   \n",
       "419  {'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...   \n",
       "420  {'GT': '0/0', 'AD': '35,0,0', 'DP': '35', 'GQ'...   \n",
       "421  {'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...   \n",
       "\n",
       "                                             B.vinkii3  \\\n",
       "0    {'GT': '0/0', 'AD': '0,0,0,0', 'DP': '0', 'GQ'...   \n",
       "1    {'GT': '0/0', 'AD': '42,0', 'DP': '42', 'GQ': ...   \n",
       "2    {'GT': '0/0', 'AD': '0,0', 'DP': '0', 'GQ': '0...   \n",
       "3    {'GT': '0/0', 'AD': '0,0', 'DP': '0', 'GQ': '0...   \n",
       "4    {'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...   \n",
       "..                                                 ...   \n",
       "417  {'GT': '0/0', 'AD': '0,0', 'DP': '0', 'GQ': '0...   \n",
       "418  {'GT': '0/0', 'AD': '0,0,0,0', 'DP': '0', 'GQ'...   \n",
       "419  {'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...   \n",
       "420  {'GT': '0/0', 'AD': '50,0,0', 'DP': '50', 'GQ'...   \n",
       "421  {'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...   \n",
       "\n",
       "                                            B.vinkii3a  \\\n",
       "0    {'GT': '0/0', 'AD': '3,0,0,0', 'DP': '3', 'GQ'...   \n",
       "1    {'GT': '0/0', 'AD': '266,0', 'DP': '266', 'GQ'...   \n",
       "2    {'GT': '0/0', 'AD': '349,0', 'DP': '349', 'GQ'...   \n",
       "3    {'GT': '0/0', 'AD': '349,0', 'DP': '349', 'GQ'...   \n",
       "4    {'GT': '0/0', 'AD': '349,0,0', 'DP': '349', 'G...   \n",
       "..                                                 ...   \n",
       "417  {'GT': '0/0', 'AD': '17,0', 'DP': '17', 'GQ': ...   \n",
       "418  {'GT': '0/0', 'AD': '17,0,0,0', 'DP': '17', 'G...   \n",
       "419  {'GT': '0/0', 'AD': '17,0,0', 'DP': '17', 'GQ'...   \n",
       "420  {'GT': '0/0', 'AD': '269,0,0', 'DP': '269', 'G...   \n",
       "421  {'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...   \n",
       "\n",
       "                                            B.vinkii3b  \\\n",
       "0    {'GT': '0/0', 'AD': '2,0,0,0', 'DP': '2', 'GQ'...   \n",
       "1    {'GT': '0/0', 'AD': '155,0', 'DP': '155', 'GQ'...   \n",
       "2    {'GT': '0/0', 'AD': '265,0', 'DP': '265', 'GQ'...   \n",
       "3    {'GT': '0/0', 'AD': '265,0', 'DP': '265', 'GQ'...   \n",
       "4    {'GT': '0/0', 'AD': '265,0,0', 'DP': '265', 'G...   \n",
       "..                                                 ...   \n",
       "417  {'GT': '0/0', 'AD': '2,0', 'DP': '2', 'GQ': '6...   \n",
       "418  {'GT': '0/0', 'AD': '2,0,0,0', 'DP': '2', 'GQ'...   \n",
       "419  {'GT': '0/0', 'AD': '2,0,0', 'DP': '2', 'GQ': ...   \n",
       "420  {'GT': '0/0', 'AD': '200,0,0', 'DP': '200', 'G...   \n",
       "421  {'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...   \n",
       "\n",
       "                                         B.weigallii57  \n",
       "0    {'GT': '0/0', 'AD': '0,0,0,0', 'DP': '0', 'GQ'...  \n",
       "1    {'GT': '0/0', 'AD': '73,0', 'DP': '73', 'GQ': ...  \n",
       "2    {'GT': '0/0', 'AD': '92,0', 'DP': '92', 'GQ': ...  \n",
       "3    {'GT': '0/0', 'AD': '92,0', 'DP': '92', 'GQ': ...  \n",
       "4    {'GT': '0/0', 'AD': '92,0,0', 'DP': '92', 'GQ'...  \n",
       "..                                                 ...  \n",
       "417  {'GT': '0/0', 'AD': '0,0', 'DP': '0', 'GQ': '0...  \n",
       "418  {'GT': '0/0', 'AD': '0,0,0,0', 'DP': '0', 'GQ'...  \n",
       "419  {'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...  \n",
       "420  {'GT': '0/0', 'AD': '106,0,0', 'DP': '106', 'G...  \n",
       "421  {'GT': '0/0', 'AD': '0,0,0', 'DP': '0', 'GQ': ...  \n",
       "\n",
       "[422 rows x 215 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c408ee88-2bde-466a-9812-ab3e4b785aae",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "4de7bfb8-b7c9-4f8e-994e-d549bafcea36",
   "metadata": {},
   "source": [
    "## 2. Extract the genotypes to make a table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "358ab638-06b5-4a45-a488-8a7ac5a1ea8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fields to keep?\n",
    "# https://gatk.broadinstitute.org/hc/en-us/articles/360035531692-VCF-Variant-Call-Format\n",
    "# https://gatk.broadinstitute.org/hc/en-us/articles/360035531672-Haplotype-map-format?utm_source=chatgpt.com \n",
    "\n",
    "# GT:AD:DP:GQ:PGT:PID:PL:PS\n",
    "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
    "##FORMAT=<ID=AD,Number=R,Type=Integer,Description=\"Allelic depths for the ref and alt alleles in the order listed\">\n",
    "##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Approximate read depth (reads with MQ=255 or with bad mates are filtered)\">\n",
    "##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">\n",
    "##FORMAT=<ID=PGT,Number=1,Type=String,Description=\"Physical phasing haplotype information, describing how the alternate alleles are phased in relation to one another\">\n",
    "##FORMAT=<ID=PID,Number=1,Type=String,Description=\"Physical phasing ID information, where each unique ID within a given sample (but not across samples) connects records within a phasing group\">\n",
    "##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">\n",
    "## PS= phase set\n",
    "\n",
    "# Example:\n",
    "# 0/0:1,0:1:3:0,3,12\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f5763ead-17d3-4fe8-bec4-2aa46ecf7c3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make the table showing genotypes of each samples for each conserved gene.\n",
    "\n",
    "def extract_genotypes_with_pos(df):\n",
    "    sample_cols = [c for c in df.columns if isinstance(df[c].iloc[0], dict)]\n",
    "\n",
    "    gt_df = pd.DataFrame({\n",
    "        sample: df[sample].apply(lambda x: x.get(\"GT\", None))\n",
    "        for sample in sample_cols\n",
    "    })\n",
    "\n",
    "    return pd.concat([df[[\"CHROM\", \"POS\"]], gt_df], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1e5fe7d9-dc6c-452e-bb1d-a04249fd7262",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>B.S.Pet.102</th>\n",
       "      <th>B.S.Pet.104</th>\n",
       "      <th>B.S.Pet.105</th>\n",
       "      <th>B.S.Pet.127</th>\n",
       "      <th>B.S.Pet.128</th>\n",
       "      <th>B.S.Pet.129</th>\n",
       "      <th>B.S.Pet.156</th>\n",
       "      <th>B.S.Pet.157</th>\n",
       "      <th>...</th>\n",
       "      <th>B.uniflora167</th>\n",
       "      <th>B.vinkii134</th>\n",
       "      <th>B.vinkii138</th>\n",
       "      <th>B.vinkii197</th>\n",
       "      <th>B.vinkii199</th>\n",
       "      <th>B.vinkii208</th>\n",
       "      <th>B.vinkii3</th>\n",
       "      <th>B.vinkii3a</th>\n",
       "      <th>B.vinkii3b</th>\n",
       "      <th>B.weigallii57</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>scaffold11</td>\n",
       "      <td>2133652</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>scaffold11</td>\n",
       "      <td>5971895</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>scaffold11</td>\n",
       "      <td>6230609</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>scaffold11</td>\n",
       "      <td>6230610</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>scaffold11</td>\n",
       "      <td>6230612</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>...</td>\n",
       "      <td>0/2</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>417</th>\n",
       "      <td>scaffold3</td>\n",
       "      <td>4691596</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>418</th>\n",
       "      <td>scaffold3</td>\n",
       "      <td>4691597</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419</th>\n",
       "      <td>scaffold3</td>\n",
       "      <td>4691599</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0|0</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>420</th>\n",
       "      <td>scaffold3</td>\n",
       "      <td>6043092</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421</th>\n",
       "      <td>scaffold15</td>\n",
       "      <td>42144934</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>...</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "      <td>0/0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>422 rows × 192 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM       POS B.S.Pet.102 B.S.Pet.104 B.S.Pet.105 B.S.Pet.127  \\\n",
       "0    scaffold11   2133652         0/0         0/0         0/0         0/0   \n",
       "1    scaffold11   5971895         0/0         0/0         0/0         0/0   \n",
       "2    scaffold11   6230609         0/0         0/0         0/0         0/0   \n",
       "3    scaffold11   6230610         0/0         0/0         0/0         0/0   \n",
       "4    scaffold11   6230612         0/0         0/0         0/0         0/0   \n",
       "..          ...       ...         ...         ...         ...         ...   \n",
       "417   scaffold3   4691596         0/0         0/0         0/0         0/0   \n",
       "418   scaffold3   4691597         0/0         0/0         0/0         0/0   \n",
       "419   scaffold3   4691599         0/0         0/0         0/0         0/0   \n",
       "420   scaffold3   6043092         0/0         0/0         0/0         0/0   \n",
       "421  scaffold15  42144934         0/0         0/0         0/0         0/0   \n",
       "\n",
       "    B.S.Pet.128 B.S.Pet.129 B.S.Pet.156 B.S.Pet.157  ... B.uniflora167  \\\n",
       "0           0/0         0/0         0/0         0/0  ...           0/0   \n",
       "1           0/0         0/0         0/0         0/0  ...           0/0   \n",
       "2           0/0         0/0         0/0         0/0  ...           0/0   \n",
       "3           0/0         0/0         0/0         0/0  ...           0/0   \n",
       "4           0/0         0/0         0/0         0/0  ...           0/2   \n",
       "..          ...         ...         ...         ...  ...           ...   \n",
       "417         0/0         0/0         0/0         0/0  ...           0/0   \n",
       "418         0/0         0/0         0/0         0/0  ...           0/0   \n",
       "419         0/0         0/0         0/0         0|0  ...           0/0   \n",
       "420         0/0         0/0         0/0         0/0  ...           0/0   \n",
       "421         0/0         0/0         0/0         0/0  ...           0/0   \n",
       "\n",
       "    B.vinkii134 B.vinkii138 B.vinkii197 B.vinkii199 B.vinkii208 B.vinkii3  \\\n",
       "0           0/0         0/0         0/0         0/0         0/0       0/0   \n",
       "1           0/0         0/0         0/0         0/0         0/0       0/0   \n",
       "2           0/0         0/0         0/0         0/0         0/0       0/0   \n",
       "3           0/0         0/0         0/0         0/0         0/0       0/0   \n",
       "4           0/0         0/0         0/0         0/0         0/0       0/0   \n",
       "..          ...         ...         ...         ...         ...       ...   \n",
       "417         0/0         0/0         0/0         0/0         0/0       0/0   \n",
       "418         0/0         0/0         0/0         0/0         0/0       0/0   \n",
       "419         0/0         0/0         0/0         0/0         0/0       0/0   \n",
       "420         0/0         0/0         0/0         0/0         0/0       0/0   \n",
       "421         0/0         0/0         0/0         0/0         0/0       0/0   \n",
       "\n",
       "    B.vinkii3a B.vinkii3b B.weigallii57  \n",
       "0          0/0        0/0           0/0  \n",
       "1          0/0        0/0           0/0  \n",
       "2          0/0        0/0           0/0  \n",
       "3          0/0        0/0           0/0  \n",
       "4          0/0        0/0           0/0  \n",
       "..         ...        ...           ...  \n",
       "417        0/0        0/0           0/0  \n",
       "418        0/0        0/0           0/0  \n",
       "419        0/0        0/0           0/0  \n",
       "420        0/0        0/0           0/0  \n",
       "421        0/0        0/0           0/0  \n",
       "\n",
       "[422 rows x 192 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check with present data\n",
    "\n",
    "extract_genotypes_with_pos(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "89761a6b-a2cb-4261-b8fe-a7ac98da3420",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's output some data to examine the dataframe\n",
    "\n",
    "# 1. Run function\n",
    "gt_table = extract_genotypes_with_pos(df)\n",
    "\n",
    "# 2. Save to CSV\n",
    "gt_table.to_csv(\"genotypes.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "267d61dd-a864-4413-b2b5-6b27aa7e7f43",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CHROM            422\n",
      "POS              422\n",
      "B.S.Pet.102        7\n",
      "B.S.Pet.104        0\n",
      "B.S.Pet.105        0\n",
      "                ... \n",
      "B.vinkii208       16\n",
      "B.vinkii3         16\n",
      "B.vinkii3a        40\n",
      "B.vinkii3b        36\n",
      "B.weigallii57     29\n",
      "Length: 192, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Example gt_table\n",
    "# gt_table = pd.DataFrame({\n",
    "#     'sample1': ['0/0', '0/1', '1/1', '0|0'],\n",
    "#     'sample2': ['0/1', '0/0', '1/1', '1|1']\n",
    "# })\n",
    "\n",
    "# Count values different from \"0/0\" and \"0|0\" in each column\n",
    "counts = (~gt_table.isin(['0/0', '0|0'])).sum()\n",
    "print(counts)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "31b35e2e-62a6-47cc-a624-3d26a566a707",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "B.S.Sym.117: ['0/0' '1/1' '0/1' '1|1' '0/2']\n",
      "B.S.Sym.117a: ['0/0' '1/1' '0/1' '1|1' '0/2']\n",
      "B.S.Sym.28: ['0/0' '0/1' '1/1' '1|1' '0|2' '1/2' '1|2' '2|2']\n",
      "B.S.Sym.28a: ['0/0' '0/1' '1/1' '1|1']\n",
      "B.S.Sym.32: ['0/0' '1/1' '0/2' '1|1' '0|1']\n",
      "B.S.Sym.48: ['0/0' '1/1' '1|1']\n",
      "B.S.Sym.48a: ['0/0' '1/1' '1|1' '2/2' '0/1']\n",
      "B.S.Sym.B32.140: ['0/0' '0/1' '1/1' '1|1' '2|2' '0/3' '3|3' '0/2' '0|2' '0|1' '1|2']\n",
      "B.S.Sym.B34.121: ['0/0' '1/1' '1|1' '2|2']\n",
      "B.S.Sym.ELAE46.209: ['0/0' '1/1' '1|1' '0|1' '0/1' '2|2' '2/2']\n"
     ]
    }
   ],
   "source": [
    "# What are the genotypes values? (shown only for B.S.Sym* samples).\n",
    "\n",
    "unique_values = {\n",
    "    col: gt_table[col].unique()\n",
    "    for col in gt_table.columns\n",
    "    if col.startswith(\"B.S.Sym\")\n",
    "}\n",
    "\n",
    "for col, values in unique_values.items():\n",
    "    print(f\"{col}: {values}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d4b592c-c12e-4661-b459-4e870df65d41",
   "metadata": {},
   "source": [
    "## 3. On another dataframe, extract the SIFT annotations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ace49499-d7d6-4bf3-8291-4852b5625e1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. Define ANN\n",
    "ann_columns = [\n",
    "    \"Allele\",\n",
    "    \"Annotation\",\n",
    "    \"Impact\",\n",
    "    \"Gene_Name\",\n",
    "    \"Gene_ID\",\n",
    "    \"Feature_Type\",\n",
    "    \"Feature_ID\",\n",
    "    \"Transcript_BioType\",\n",
    "    \"Rank\",\n",
    "    \"HGVS_c\",\n",
    "    \"HGVS_p\",\n",
    "    \"cDNA_pos\",\n",
    "    \"CDS_pos\",\n",
    "    \"AA_pos\",\n",
    "    \"Distance\",\n",
    "    \"Errors\"\n",
    "]\n",
    "\n",
    "# Define the function\n",
    "#\n",
    "def parse_ann_field(ann_string):\n",
    "    if pd.isna(ann_string):\n",
    "        return pd.Series([None] * len(ann_columns))\n",
    "    \n",
    "    # Take first annotation if multiple\n",
    "    first_ann = ann_string.split(\",\")[0]\n",
    "    \n",
    "    values = first_ann.split(\"|\")\n",
    "    \n",
    "    # Pad if missing fields\n",
    "    if len(values) < len(ann_columns):\n",
    "        values += [None] * (len(ann_columns) - len(values))\n",
    "    \n",
    "    return pd.Series(values[:len(ann_columns)])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ea5bb380-1426-4bc4-b388-e4dd800b9a5a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Allele</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Impact</th>\n",
       "      <th>Gene_Name</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Feature_Type</th>\n",
       "      <th>Feature_ID</th>\n",
       "      <th>Transcript_BioType</th>\n",
       "      <th>Rank</th>\n",
       "      <th>HGVS_c</th>\n",
       "      <th>HGVS_p</th>\n",
       "      <th>cDNA_pos</th>\n",
       "      <th>CDS_pos</th>\n",
       "      <th>AA_pos</th>\n",
       "      <th>Distance</th>\n",
       "      <th>Errors</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>upstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>Bma030423</td>\n",
       "      <td>null.4442</td>\n",
       "      <td>transcript</td>\n",
       "      <td>Bma030423</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td></td>\n",
       "      <td>c.-634G&gt;A</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>634</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>Bma005962.1</td>\n",
       "      <td>null.4683</td>\n",
       "      <td>transcript</td>\n",
       "      <td>Bma005962.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>4/4</td>\n",
       "      <td>c.934A&gt;T</td>\n",
       "      <td>p.Thr312Ser</td>\n",
       "      <td>934/1413</td>\n",
       "      <td>934/1413</td>\n",
       "      <td>312/470</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>Bma005980.2</td>\n",
       "      <td>null.4702</td>\n",
       "      <td>transcript</td>\n",
       "      <td>Bma005980.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>7/10</td>\n",
       "      <td>c.822T&gt;G</td>\n",
       "      <td>p.Asn274Lys</td>\n",
       "      <td>822/1416</td>\n",
       "      <td>822/1416</td>\n",
       "      <td>274/471</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>MODERATE</td>\n",
       "      <td>Bma005980.2</td>\n",
       "      <td>null.4702</td>\n",
       "      <td>transcript</td>\n",
       "      <td>Bma005980.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>7/10</td>\n",
       "      <td>c.821A&gt;G</td>\n",
       "      <td>p.Asn274Ser</td>\n",
       "      <td>821/1416</td>\n",
       "      <td>821/1416</td>\n",
       "      <td>274/471</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GACCTAC</td>\n",
       "      <td>stop_gained&amp;disruptive_inframe_insertion</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Bma005980.2</td>\n",
       "      <td>null.4702</td>\n",
       "      <td>transcript</td>\n",
       "      <td>Bma005980.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>7/10</td>\n",
       "      <td>c.818_819insGTAGGT</td>\n",
       "      <td>p.Val273_Asn274insTerVal</td>\n",
       "      <td>818/1416</td>\n",
       "      <td>818/1416</td>\n",
       "      <td>273/471</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>417</th>\n",
       "      <td>T</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>Bma016853.1</td>\n",
       "      <td>null.13036</td>\n",
       "      <td>transcript</td>\n",
       "      <td>Bma016853.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td></td>\n",
       "      <td>c.*3434_*3436delATC</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>3434</td>\n",
       "      <td>WARNING_TRANSCRIPT_NO_START_CODON</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>418</th>\n",
       "      <td>ACTCAAAACATGAAGACATG</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>Bma016853.1</td>\n",
       "      <td>null.13036</td>\n",
       "      <td>transcript</td>\n",
       "      <td>Bma016853.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td></td>\n",
       "      <td>c.*3435delT</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>3435</td>\n",
       "      <td>WARNING_TRANSCRIPT_NO_START_CODON</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>419</th>\n",
       "      <td>T</td>\n",
       "      <td>downstream_gene_variant</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>Bma016853.1</td>\n",
       "      <td>null.13036</td>\n",
       "      <td>transcript</td>\n",
       "      <td>Bma016853.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td></td>\n",
       "      <td>c.*3436C&gt;T</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>3436</td>\n",
       "      <td>WARNING_TRANSCRIPT_NO_START_CODON</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>420</th>\n",
       "      <td>A</td>\n",
       "      <td>intergenic_region</td>\n",
       "      <td>MODIFIER</td>\n",
       "      <td>Bma030990-Bma030991</td>\n",
       "      <td>null.13167-null.13168</td>\n",
       "      <td>intergenic_region</td>\n",
       "      <td>null.13167-null.13168</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>n.6043092T&gt;A</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421</th>\n",
       "      <td>A</td>\n",
       "      <td>frameshift_variant</td>\n",
       "      <td>HIGH</td>\n",
       "      <td>Bma014354.2</td>\n",
       "      <td>null.10970</td>\n",
       "      <td>transcript</td>\n",
       "      <td>Bma014354.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>5/7</td>\n",
       "      <td>c.1601delC</td>\n",
       "      <td>p.Thr534fs</td>\n",
       "      <td>1601/2226</td>\n",
       "      <td>1601/2226</td>\n",
       "      <td>534/741</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>422 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Allele                                Annotation    Impact  \\\n",
       "0                       A                     upstream_gene_variant  MODIFIER   \n",
       "1                       A                          missense_variant  MODERATE   \n",
       "2                       C                          missense_variant  MODERATE   \n",
       "3                       C                          missense_variant  MODERATE   \n",
       "4                 GACCTAC  stop_gained&disruptive_inframe_insertion      HIGH   \n",
       "..                    ...                                       ...       ...   \n",
       "417                     T                   downstream_gene_variant  MODIFIER   \n",
       "418  ACTCAAAACATGAAGACATG                   downstream_gene_variant  MODIFIER   \n",
       "419                     T                   downstream_gene_variant  MODIFIER   \n",
       "420                     A                         intergenic_region  MODIFIER   \n",
       "421                     A                        frameshift_variant      HIGH   \n",
       "\n",
       "               Gene_Name                Gene_ID       Feature_Type  \\\n",
       "0              Bma030423              null.4442         transcript   \n",
       "1            Bma005962.1              null.4683         transcript   \n",
       "2            Bma005980.2              null.4702         transcript   \n",
       "3            Bma005980.2              null.4702         transcript   \n",
       "4            Bma005980.2              null.4702         transcript   \n",
       "..                   ...                    ...                ...   \n",
       "417          Bma016853.1             null.13036         transcript   \n",
       "418          Bma016853.1             null.13036         transcript   \n",
       "419          Bma016853.1             null.13036         transcript   \n",
       "420  Bma030990-Bma030991  null.13167-null.13168  intergenic_region   \n",
       "421          Bma014354.2             null.10970         transcript   \n",
       "\n",
       "                Feature_ID Transcript_BioType  Rank               HGVS_c  \\\n",
       "0                Bma030423     protein_coding                  c.-634G>A   \n",
       "1              Bma005962.1     protein_coding   4/4             c.934A>T   \n",
       "2              Bma005980.2     protein_coding  7/10             c.822T>G   \n",
       "3              Bma005980.2     protein_coding  7/10             c.821A>G   \n",
       "4              Bma005980.2     protein_coding  7/10   c.818_819insGTAGGT   \n",
       "..                     ...                ...   ...                  ...   \n",
       "417            Bma016853.1     protein_coding        c.*3434_*3436delATC   \n",
       "418            Bma016853.1     protein_coding                c.*3435delT   \n",
       "419            Bma016853.1     protein_coding                 c.*3436C>T   \n",
       "420  null.13167-null.13168                                  n.6043092T>A   \n",
       "421            Bma014354.2     protein_coding   5/7           c.1601delC   \n",
       "\n",
       "                       HGVS_p   cDNA_pos    CDS_pos   AA_pos Distance  \\\n",
       "0                                                                 634   \n",
       "1                 p.Thr312Ser   934/1413   934/1413  312/470            \n",
       "2                 p.Asn274Lys   822/1416   822/1416  274/471            \n",
       "3                 p.Asn274Ser   821/1416   821/1416  274/471            \n",
       "4    p.Val273_Asn274insTerVal   818/1416   818/1416  273/471            \n",
       "..                        ...        ...        ...      ...      ...   \n",
       "417                                                              3434   \n",
       "418                                                              3435   \n",
       "419                                                              3436   \n",
       "420                                                                     \n",
       "421                p.Thr534fs  1601/2226  1601/2226  534/741            \n",
       "\n",
       "                                Errors  \n",
       "0                                       \n",
       "1                                       \n",
       "2                                       \n",
       "3                                       \n",
       "4                                       \n",
       "..                                 ...  \n",
       "417  WARNING_TRANSCRIPT_NO_START_CODON  \n",
       "418  WARNING_TRANSCRIPT_NO_START_CODON  \n",
       "419  WARNING_TRANSCRIPT_NO_START_CODON  \n",
       "420                                     \n",
       "421                                     \n",
       "\n",
       "[422 rows x 16 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ann_df = df[\"ANN\"].apply(parse_ann_field)\n",
    "ann_df.columns = ann_columns\n",
    "ann_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "442887cd-bec3-42d7-8287-d7fb0d73ea89",
   "metadata": {},
   "source": [
    "## 4. Discriminate the High-impact variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "684e6c2d-e13c-4ef8-9c5e-48d69b89fc35",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The plan: identify deleterious alleles and link them to specific samples.\n",
    "\n",
    "# 1. Detect deleterious alleles:\n",
    "# Best options to pin down deleterious alleles:\n",
    "# A. ALT allele\n",
    "# B. high conservation score (GERP>2).\n",
    "# C. Functional effect (missense, LOF).\n",
    "\n",
    "# 2. Convert genotypes into deleterious alleles count\n",
    "# GT\tdeleterious alleles\n",
    "# 0/0\t0\n",
    "# 0/1\t1\n",
    "# 1/1\t2\n",
    "# ./ .\tNA\n",
    "\n",
    "# 3. Compute deleterious load per sample. \n",
    "# 4. Identify which variants are deleterious.\n",
    "# 5. Visualize results (stripplot, load per sample)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de19c89f-3d50-4bda-a7c0-37562fc2b411",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c49ea524-aece-4112-9c15-2cec9d850951",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "df4d9659-ca14-4eb8-9977-93604378d2e9",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'vcf_an' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[18], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# This will be shown in the column below\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m \u001b[43mvcf_an\u001b[49m[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mAnnotation_Impact\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n",
      "\u001b[0;31mNameError\u001b[0m: name 'vcf_an' is not defined"
     ]
    }
   ],
   "source": [
    "# This will be shown in the column below\n",
    "vcf_an[\"Annotation_Impact\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34c5d459-cf73-4ac1-97d5-7d09a7fb8fe8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This will be shown in the column below\n",
    "vcf_an[\"Annotation_Impact\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "682a4680-6248-4be4-8a70-47bb577e6e4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This will be shown in the column below\n",
    "vcf_an[\"Annotation_Impact\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a4c0e12-4194-4175-907e-29297826e633",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This will be shown in the column below\n",
    "vcf_an[\"Annotation_Impact\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81c68d14-3a0f-49e1-9c13-9334ce3a41c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Number of distinct (unique) values\n",
    "# --------------------------------------------------------------\n",
    "n_unique = vcf_an[\"Annotation_Impact\"].nunique(dropna=True)\n",
    "print(f\"Number of distinct Annotation_Impact values (excluding NaN): {n_unique}\")\n",
    "\n",
    "# Frequency table – each value and its count\n",
    "# --------------------------------------------------------------\n",
    "freq_tbl = vcf_an[\"Annotation_Impact\"].value_counts(dropna=False)  # keep NaN as a category if you like\n",
    "print(\"\\nFrequency of each Annotation_Impact value:\")\n",
    "print(freq_tbl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "257b38f0-7fa2-43e5-86e6-b360046e0257",
   "metadata": {},
   "outputs": [],
   "source": [
    "# We write the DataFrame to a CSV file to be able to check all the columns (comma‑separated, with a header line)\n",
    "vcf_an.to_csv(\"vcf_an.csv\", index=False)   # index=False omits the pandas row index\n",
    "print(\"CSV written to vcf_an.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a5fe372-3cff-4414-bf03-6df444928c78",
   "metadata": {},
   "source": [
    "### Selecting high-effect variants "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fce4d98-a307-499b-8ef9-2164aa49f368",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select the high-impact deleterious alleles variants\n",
    "vcf_an[vcf_an[\"Annotation_Impact\"]==\"HIGH\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "439966c0-def8-427c-8708-ee7eeb91f662",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
